### PR TITLE
MessageProcessingAbortedExcepton made public

### PR DESCRIPTION
**Pull request by tmasternak**
This exception needs to be exposed so that v6 transport seam can properly support message receive rollback due to recoverability functionality. For example #3216.
This connects to #3078 spike for using OperationAbortedException instead. In addition it connect also to #3044.

### CODE REVIEW

**Comment by tmasternak** [comment]
>I wouldn't say just recoverability, how about "thrown to request the message to be rolled back to the input queue" ?

@andreasohlund This touches on an interesting aspect. I think that the type should be public but the constructor should be available only to core. If any other behavior will be able to throw this it will:
* Work as expected if new behavior will be before FLR in the pipeline 
* Treated as any other error if new behavior will be placed after FLR

[See](https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Recoverability/FirstLevelRetries/FirstLevelRetriesBehavior.cs#L30)

**Comment by tmasternak** [comment]
@SimonCropp why do we need public exceptions to be serializable and have protected constructor?

**Comment by andreasohlund** [comment]
>I think that the type should be public but the constructor should be available only to core

That would make it hard for the transports to unit test rollback behavior?

**Comment by tmasternak** [comment]
Good point. How about internal?

**Comment by timbussmann** [comment]
> Good point. How about internal?

Doesn't change anything for user written behaviors. So I'm :-1: for having public classes with non public ctors.

**Comment by tmasternak** [comment]
@timbussmann so maybe changing FLR is a way to go here?

**Comment by timbussmann** [comment]
I guess it shows a design flaw with an exception which is related to a specific behavior and needs to be handled by other behaviors and even the transport. But I guess that's a different story. 
I'd prefer a public ctor even with the risk of users throwing that exception.

**Comment by tmasternak** [comment]
I agree. And just to add it's not an exceptional situation that we are modelling (see #3044)

File: src/NServiceBus.Core/Pipeline/Incoming/MessageProcessingAbortedException.cs [hunk 1]
```diff
@@ -1,8 +1,30 @@
...
 {
     using System;
+    using System.Runtime.Serialization;
+    using JetBrains.Annotations;
+    using NServiceBus.Features;
+    using NServiceBus.Transports;
 
-    class MessageProcessingAbortedException : Exception
+    /// <summary>
+    /// This exception can be thrown by any of recoverability behaviors ie.e <see cref="FirstLevelRetries"/>, <see cref="SecondLevelRetries"/> or <see cref="StoreFaultsInErrorQueue"/>
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Pipeline/Incoming/MessageProcessingAbortedException.cs]: Still don't see the value of mentioning the behaviors or even recoverability at all. The doco is there to tell the transport implementor what he should do?

I.e. roll the message back to the queue and do not trigger the circuit breaker?

This made me realize that we should add this to the `NServiceBus.Transports` namespace

File: src/NServiceBus.Core/Pipeline/Incoming/MessageProcessingAbortedException.cs [hunk 2]
```diff
@@ -1,8 +1,30 @@
 namespace NServiceBus
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Pipeline/Incoming/MessageProcessingAbortedException.cs]: Propose we change namespace, see below for motivation

**Comment by johnsimons** [comment]
> I guess it shows a design flaw with an exception which is related to a specific behavior and needs to be handled by other behaviors and even the transport.

Totally agree with @timbussmann on this, expecting the transport implementers to catch this exception and deal with it is flaw in many ways.
Maybe in the Java world would work because exceptions are declared upfront but in .net it does not.
And we can document the hell out of it and I can almost guarantee someone will forget :disappointed: 

So what other proposals were put forward to deal with this situation ?

**Comment by SimonCropp** [comment]
@tmasternak exceptions need to have the serializable constructor so they can cross appdomain boundaries

File: src/NServiceBus.Core/Pipeline/Incoming/MessageProcessingAbortedException.cs (hunk 1 above)
Comment (tmasternak) [inline src/NServiceBus.Core/Pipeline/Incoming/MessageProcessingAbortedException.cs]: :+1:

**Comment by tmasternak** [comment]
I've made FLR changes to make pipeline behavior consistent.

**Comment by andreasohlund** [comment]
So give that we're going ahead with #3242 this one is obsolete right?

**Comment by tmasternak** [comment]
correct. closing

**Pull request closed without merging**